### PR TITLE
vtgate batch API: core functionality

### PR DESCRIPTION
**Pull request by sougou**

### CODE REVIEW

**Comment by alainjobart** [comment]
LGTM

File: go/vt/vtgate/resolver.go [hunk 1]
```diff
@@ -235,51 +236,37 @@ func (res *Resolver) ExecuteBatchKeyspaceIds(ctx context.Context, query *proto.K
...
-				keyspace = newKeyspace
-				resharding = true
-			}
-			// check shards change for horizontal resharding
-			if !StrsEquals(newShards, shards) {
-				shards = newShards
-				resharding = true
+			// Use reflect to see if the request has changed.
+			if reflect.DeepEqual(*batchRequest, *newBatchRequest) {
+				return qrs, err
```
Comment (yaoshengzhe) [inline go/vt/vtgate/resolver.go]: why do we reuse the "qrs" instead of re-sending the scatter requests?

Comment (sougou) [inline go/vt/vtgate/resolver.go]: Note that this entire block of code is error handling for a RETRY error code.
Retries already happen at the shard_conn level. But a retry at shard_conn level will always fail if there was a resharding event. So, at this level we perform a re-resolve and retry. In this case, if the newBatchRequest has not changed after a re-resolve, then we know that we should not retry any more. So, we just return whatever shard_conn returned.

Comment (yaoshengzhe) [inline go/vt/vtgate/resolver.go]: Got it. So essentially error code is used here to determine the system behavior. Could we have proper documentation for that ?

Comment (sougou) [inline go/vt/vtgate/resolver.go]: Agree. I'll add comments on the next change.

**Pull request merged**